### PR TITLE
Use embedded title for other track types

### PR DESCRIPTION
**Pull request by ThatNerdyPikachu**
**Changes**
This adjusts the media stream DisplayTitle logic to use a switch, and makes it so every track type uses its embedded title, if one exists.

**Issues**
Implements https://features.jellyfin.org/posts/385/prefer-embdedded-titles-for-audio-subtitle-tracks

### CODE REVIEW

**Comment by ZadenRB** [comment]
Related to #2773, if you want to take a look at my changes to how the languages are fetched in that PR and implement those here I can close that PR and just have it included in this one.

**Comment by ThatNerdyPikachu** [comment]
@ZadenRB Sure thing.

**Comment by ThatNerdyPikachu** [comment]
@ZadenRB How's that?

**Comment by ZadenRB** [comment]
@ThatNerdyPikachu Good, just missing the code to get the full language for subtitles as well.

**Comment by ThatNerdyPikachu** [comment]
@ZadenRB Was just fixing as you commented that :)

**Comment by ZadenRB** [comment]
Looks solid, I'll close my PR. If you have any thoughts on the discussion on that PR about the tag filter potentially being too broad, feel free to make changes or just comment them here.

**Comment by ThatNerdyPikachu** [comment]
Sounds good.

**Comment by ThatNerdyPikachu** [comment]
@ZadenRB For some reason, this isn't turning ``Eng`` to ``English``

https://e-six-two.one/i/bcn63w4s.png

Mediainfo:
```
Audio #1
ID                                       : 2
Format                                   : AC-3
Format/Info                              : Audio Coding 3
Commercial name                          : Dolby Digital
Codec ID                                 : A_AC3
Duration                                 : 21 min 20 s
Bit rate mode                            : Constant
Bit rate                                 : 384 kb/s
Channel(s)                               : 6 channels
Channel layout                           : L R C LFE Ls Rs
Sampling rate                            : 48.0 kHz
Frame rate                               : 31.250 FPS (1536 SPF)
Compression mode                         : Lossy
Stream size                              : 58.6 MiB (7%)
Language                                 : English
Service kind                             : Complete Main
Default                                  : Yes
Forced                                   : No
```

Why is that?

**Comment by ZadenRB** [comment]
@ThatNerdyPikachu Try switching `CultureTypes.NeutralCultures` on lines 90 & 168 to `CultureTypes.AllCultures`. That's a change I made to try and speed up the check but it may have broken certain languages.

**Comment by ZadenRB** [comment]
@ThatNerdyPikachu Just did some digging, it looks like `NeutralCultures` is the way we would want to go, since that will only pull in cultures associated with a language, and doing `AllCultures` pulls in like 40 varieties of English, which doesn't apply to subtitles or audio tracks. But the current code doesn't localize based on anything, so it will always be the English name, even if the UI is set to something else, and clearly has some cases where it doesn't work. If you want, you can remove the language patch or just leave it as `AllCultures` for now, and I'll open a second PR sorting out the details of the language tag.

**Comment by ThatNerdyPikachu** [comment]
@ZadenRB I'll keep the patch in, since then we don't have to open a third patch once you sort out that.

:) Thanks for looking into it.

**Comment by Artiume** [comment]
https://github.com/jellyfin/jellyfin/issues/2947#issuecomment-617531845

So I was hoping this PR would fix this person's issue since it's relatively close, but it didn't work. All of his audio tracks are marked as undefined, this is causing the track preference to always reset. Would it be feasible to add in support for their situation as well?

**Comment by ThatNerdyPikachu** [comment]
@Artiume I can take a look at their issue, provided a MediaInfo.
Will comment there.

**Comment by ThatNerdyPikachu** [comment]
Could we get this merged for 10.6.0? (I'm aware it's in feature-freeze, but this was PR'd much before that...)

File: MediaBrowser.Model/Entities/MediaStream.cs [hunk 1]
```diff
@@ -112,107 +112,131 @@ public string DisplayTitle
...
+                        }
+
+                        if (!string.IsNullOrEmpty(Title))
+                        {
+                            return attributes.AsEnumerable()
+                                // keep Tags that are not already in Title
+                                .Where(tag => Title.IndexOf(tag, StringComparison.OrdinalIgnoreCase) == -1)
+                                // attributes concatenation, starting with Title
+                                .Aggregate(new StringBuilder(Title), (builder, attr) => builder.Append(" - ").Append(attr))
+                                .ToString();
```
Comment (crobibero) [inline MediaBrowser.Model/Entities/MediaStream.cs:231]:
```suggestion
                           var result = new StringBuilder(Title);
                            foreach (var tag in attributes)
                            {
                                // Keep Tags that are not already in Title.
                                if (Title.IndexOf(tag, StringComparison.OrdinalIgnoreCase) == -1)
                                {
                                    result.Append(" - ").Append(tag);
                                }
                            }

                            return result.ToString();
```

Prefer to not use Linq. Same for Movie and Subtitle

File: MediaBrowser.Model/Entities/MediaStream.cs [hunk 2]
```diff
@@ -77,106 +77,131 @@ public string DisplayTitle
...
+                            attributes.Add(AudioCodec.GetFriendlyName(Codec));
+                        }
+                        else if (!string.IsNullOrEmpty(Profile) && !string.Equals(Profile, "lc", StringComparison.OrdinalIgnoreCase))
+                        {
+                            attributes.Add(Profile);
+                        }
+
+                        if (!string.IsNullOrEmpty(ChannelLayout))
+                        {
+                            attributes.Add(StringHelper.FirstToUpper(ChannelLayout));
```
[the remark this answers is not in the document]
Comment (ThatNerdyPikachu) [reply MediaBrowser.Model/Entities/MediaStream.cs:142]: I actually have no clue, I ported that from @ZadenRB's PR, maybe they could chime in?

File: MediaBrowser.Model/Entities/MediaStream.cs [hunk 3]
```diff
@@ -112,107 +112,131 @@ public string DisplayTitle
...
+                        }
+
+                        if (!string.IsNullOrEmpty(Title))
+                        {
+                            return attributes.AsEnumerable()
+                                // keep Tags that are not already in Title
+                                .Where(tag => Title.IndexOf(tag, StringComparison.OrdinalIgnoreCase) == -1)
+                                // attributes concatenation, starting with Title
+                                .Aggregate(new StringBuilder(Title), (builder, attr) => builder.Append(" - ").Append(attr))
+                                .ToString();
```
Comment (crobibero) [inline MediaBrowser.Model/Entities/MediaStream.cs:190]:
```suggestion
                            var result = new StringBuilder(Title);
                            foreach (var tag in attributes)
                            {
                                // Keep Tags that are not already in Title.
                                if (Title.IndexOf(tag, StringComparison.OrdinalIgnoreCase) == -1)
                                {
                                    result.Append(" - ").Append(tag);
                                }
                            }

                            return result.ToString();
```

File: MediaBrowser.Model/Entities/MediaStream.cs [hunk 4]
```diff
@@ -112,107 +112,131 @@ public string DisplayTitle
...
```
Comment (crobibero) [inline MediaBrowser.Model/Entities/MediaStream.cs:161]:
```suggestion
                            var result = new StringBuilder(Title);
                            foreach (var tag in attributes)
                            {
                                // Keep Tags that are not already in Title.
                                if (Title.IndexOf(tag, StringComparison.OrdinalIgnoreCase) == -1)
                                {
                                    result.Append(" - ").Append(tag);
                                }
                            }

                            return result.ToString();
```

File: MediaBrowser.Model/Entities/MediaStream.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (crobibero) [reply MediaBrowser.Model/Entities/MediaStream.cs:142]: This change seems fine to me, IIRC channel layouts are normally `[0-9].[0-9]`

[the remark this answers is not in the document]
Comment (ZadenRB) [reply MediaBrowser.Model/Entities/MediaStream.cs:142]: I did it for visuals on the frontend with stereo layout. To me it looks better " - Stereo" vs "- stereo" when everything else is usually title-cased, like the title of the audio track. Totally open to changing it though if others disagree 👍

File: MediaBrowser.Model/Entities/MediaStream.cs [hunk 5]
```diff
@@ -112,107 +112,146 @@ public string DisplayTitle
...
+                                if (Title.IndexOf(tag, StringComparison.OrdinalIgnoreCase) == -1)
+                                {
+                                    result.Append(" - ").Append(tag);
+                                }
+                            }
+
+                            return result.ToString();
+                        }
+
+                        return string.Join(" ", attributes);
```
Comment (crobibero) [inline MediaBrowser.Model/Entities/MediaStream.cs:203]:
```suggestion
                        return string.Join(' ', attributes);
```

**Pull request merged**